### ISSUE TITLE
what if model has conditional properties?

### ISSUE THREAD

**Issue by smorandi** [gparlakov/forms-typed#64]
assume the following:

```
interface Model {
 validFrom: string,
  validTo?: string
}
```

so i tried to do the following and ran into multiple issues i am not sure how to deal with, or rather what could be done about it for a proper solution?

```
interface Model {
  validFrom: string,
  validTo?: string
}

const factory = (initialModel: Model) => typedFormGroup<Model>({
  validFrom: typedFormControl(initialModel.validFrom),


  // does not work...?
  // validTo: typedFormControl(initialModel.validTo),

  // this works...
  validTo: typedFormControl<string | undefined>(initialModel.validTo),

});

const mfg = factory({validFrom: 'foo', validTo: 'bar'});

// getting error on controls.validTo, since it can be undefined???
combineLatest([mfg.controls.validFrom.valueChanges, mfg.controls.validTo.valueChanges]).pipe(
  //something...
).subscribe();

```
can you tell me if i am doing something wrong?

**Comment by gparlakov** [comment]
I guess it can be created with a form control instance for Id without a
value (undefined).

**Comment by smorandi** [comment]
well yes, but the point is that a validTo controller is always available, it should never be undefined. maybe when creating the types they should omit any conditionals first.

like this maybe https://www.typescriptlang.org/docs/handbook/2/mapped-types.html

removing the optional types for when you create a type of the model mapped to the individual controllers…
```
// Removes 'optional' attributes from a type's properties
type Concrete<Type> = {
  [Property in keyof Type]-?: Type[Property];
};
```

just an idea though…

**Comment by gparlakov** [comment]
Hey, @smorandi it seems I've misunderstood your initial concern. 

If I understand, the `validTo` property of the `form.controls` should not be undefined if we've defined it in the `typedFormGroup` function call - right? The following should not break:

![image](https://user-images.githubusercontent.com/3482199/140866000-ee2cd091-90b8-4885-b1b5-60f53f8f2b17.png)


It seems there is a utility type coming from TS itself that can help: `Required`:

```ts
interface ModelWithPartial {
  validFrom: string;
  validTo?: string;
}

// const factory = (initialModel: ModelWithPartial) => typedFormGroup<ModelWithPartial>({
const factory = (initialModel: ModelWithPartial) => typedFormGroup<Required<ModelWithPartial>>({
  validFrom: typedFormControl(initialModel.validFrom),


  // does not work...?
  // validTo: typedFormControl(initialModel.validTo),

  // this works...
  validTo: typedFormControl(initialModel.validTo),

});

const mfg = factory({validFrom: 'foo', validTo: 'bar'});

// getting error on controls.validTo, since it can be undefined???
// not getting an error any more
mfg.controls.validTo.valueChanges.subscribe(v => v?.includes('t'));
```


-------------------
old, misunderstood-question answer:

Hey, @smorandi that looks like it might remove the optional properties in general. So when you try to actually use them in the form (or another form) you'll not get the strong-type. 

Edit: I looked into that type and it does not remove the properties, it removes the optionality:
![image](https://user-images.githubusercontent.com/3482199/140633995-92f8eb11-392b-4f74-80e5-c78a1b405d2c.png)


Still this `-?` trick was new to me so - thanks 👍🏻 


My understanding is that the idea is to have the `typedFormGroup` function's parameter skip some of the `formControl`-s in its initializer, like so:
  
![image](https://user-images.githubusercontent.com/3482199/140634093-538e713a-3aff-462b-86ea-23399c6fcbae.png)

That would cause the issue I mentioned above - wherever one uses the optional props (like `names` and `age` in the above case) they'd have to check for `null/undefined`, which might be awkward - especially in a template.
